### DISCUSSION TITLE
Noob question warning. NVIDIA vs AMD

### DISCUSSION THREAD

**Discussion by jjlynn27** [trackmania-rl/tmrl#82] [General]
I quite enjoy training cars on trackmania using this framework, even I don't have a scooby of how it all works. I currently have amd 6600 XT GPU. Would changing this to NVIDIA 4060 improve speed of training?

Thanks.

**Comment by yannbouteiller** [comment]
Hi, I don't think you'll see a great performance boost with the 4060 as it is basically on-par with the 6600-XT. NVIDIA cards tend to work better than AMD for deep learning in general because deep learning frameworks are primarily built around CUDA but PyTorch (which is used by TMRL) now supports AMD's ROCm apparently, so if it works with your AMD card I would not switch to a 4060 just for this purpose. Cards like 4070/4080/4090 make more sense as they would enable you to train faster with larger batch sizes.

**Reply by jjlynn27** [reply]
Hi,
First of all thanks very much for sharing this framework, even if I don't understand it it's sooo much fun. 
As far as ROCm is concerned, that doesn't really work even on linux and it's completely unavailable for windows. Throw my noobines into the equation and basically it's no go for AMD. 
In that config.json files, I disabled cuda, because when I do the test, it shows 'cpu'. So it's just using CPU, which in my case is AMD 5600X (I think). 

Thanks again for your time and for sharing all this work. It's absolutely  mind boggling. I've now switched to LIDAR training but it's slow, on a very short track with just one turn and two straits it's taking more than a day to get to the end. :) (currently epoch 1481/10000).
Anyway, Happy New Year and all the best to you.

**Reply by yannbouteiller** [reply]
Thanks, it is always a pleasure to read positive feedback :)

I see, in this case, then yes the 4060 will change your life. Training on CPU in not really feasible, even with the LIDAR observations. With LIDARs, you can try to increase LR_ACTOR and LR_CRITIC in `config.json` to accelerate training a bit, as the default values are tuned for images mostly. I would try multiplying both by 10, maybe even 100 (just bear in mind that if you set these learning rates too high training will become unstable and fail).

**Reply by jjlynn27** [reply]
Yann, 
OT, I'm holding you personally responsible for me spending £300 and getting 4060. My wife was confused as to why I needed a graphics card when I already had one, to which answer was, 'Imelda, do you want me to open your shoes and bags cupboards (yes, plural). 
Anyhow both cards and your software is soo much fun, I still have no idea how it all works but it's fun finding it out. 
I've reinstalled your software and noticed that config.json is completely different, so I don't know why training is SOOO much faster. Is it the card, or configuration/software? 
One thing that I noticed, is that there is a slight hesitation when the green light for a start goes on. Also, the best times were achieved after starting server/trainer/worker. I'm thinking that it could be that I recorded rewards with the delay at the start (as I'm still not certain of the exact sequence of when to press e and when to start driving. 
It's 5am, and this is so addictive and fun.

**Reply by yannbouteiller** [reply]
Haha, well the 4060 is probably the main reason why it is faster, but I think I have also changed the default hyperparameters in `config.json`, this may help :)

The short delay at the start of episodes is intended, in fact there is a value in `config.json` to set this delay (1.5 or 2.5 by default I believe). The reason is that we don't want the AI to be collecting training samples before the light turns green (this would be hard to handle properly from a reinforcement learning perspective, as the AI wouldn't understand why the car doesn't want to move during this phase), so instead it sleeps and starts controlling the car a short while after the race starts. For testing a trained policy, you can set this delay to 0.0, but for training you should leave it to the default.

**Reply by jjlynn27** [reply]
If you are ever in mid-England, and you are up to it, I'd like to buy you a drink. You provided so much fun that I've decided to call work today and tell them that I have more important things to do. 
In the meantime, I found that param (sleep_time_at_reset) and before your post I've already changed it to 0.5 (from 1.5 default), but now I'm going to turn it back as your answer kind of makes sense in that tiny brain of mine. YANQ (yet-another-noob-question) How do I change from training to testing trained policy?

**Reply by jjlynn27** [reply]
Also, do you mind if I post a short video on YouTube (obviously with credits) of about 100 ghosts running at the same time?

**Reply by yannbouteiller** [reply]
I do not mind at all, no problem.

To train the policy:

`python -m tmrl --worker`

To test it:

`python -m tmrl --test`

**Comment by NDR008** [comment]
How do you make those videos with multiple ghosts?

**Reply by jjlynn27** [reply]
Edit replays from Create in Trackmania itself.

**Reply by yannbouteiller** [reply]
In the new version of `tmrl`, there is a `save_replay` option in `config.json` that automatically saves all ghosts.

**Reply by jjlynn27** [reply]
yep. That's the one that I used.
Yann, if you have any objections to video for whatever reason. I'll take it down.

**Reply by yannbouteiller** [reply]
> yep. That's the one that I used.
> Yann, if you have any objections to video for whatever reason. I'll take it down.
> 

I don't see why I would have any objections, don't worry. This is an MIT license library, people can use it however they want.
